src/utils/getActionContext.ts

```diff
@@ -0,0 +1,61 @@
+import type { Input } from 'hono'
+import type { ActionContext, Context } from '../types/context.js'
+import type { Env } from '../types/env.js'
+
+type GetActionContextParameters<
+  env extends Env = Env,
+  path extends string = string,
+  input extends Input = {},
+> = {
+  context: Context<env, path, input>
+}
+
+type GetActionContextReturnType<
+  env extends Env = Env,
+  path extends string = string,
+  input extends Input = {},
+> = {
+  context: ActionContext<env, path, input>
+}
+
+export function getActionContext<
+  env extends Env,
+  path extends string,
+  input extends Input = {},
+>(
+  parameters: GetActionContextParameters<env, path, input>,
+): GetActionContextReturnType<env, path, input> {
+  const { context } = parameters
+  const { env, frameData, req, verified, origin } = context || {}
+
+  if (!frameData)
+    throw new Error('Frame data must be present for action handlers.')
+
+  return {
+    context: {
+      env,
+      actionData: {
+        buttonIndex: 1,
+        castId: frameData.castId,
+        fid: frameData.fid,
+        network: frameData.network,
+        messageHash: frameData.messageHash,
+        timestamp: frameData.timestamp,
+        url: frameData.url,
+      },
+      error: ({ message, statusCode }) => ({
+        data: { message, statusCode },
+        format: 'action',
+      }),
+      message: (message) => ({
+        data: { message, statusCode: 200 },
+        format: 'action',
```

Comment:
Maybe we don't need this, and just use `c.res`?

Reply:
> Maybe we don't need this, and just use `c.res`?

agree, maybe I've overthought this API.

Reply:
refactored in 9a16f47